db/db_impl/db_impl.cc

```diff
@@ -678,36 +680,20 @@ void DBImpl::PrintStatistics() {
   }
 }
 
-void DBImpl::StartTimedTasks() {
-  unsigned int stats_dump_period_sec = 0;
-  unsigned int stats_persist_period_sec = 0;
+void DBImpl::StartStatsDumpScheduler() {
+#ifndef ROCKSDB_LITE
   {
     InstrumentedMutexLock l(&mutex_);
-    stats_dump_period_sec = mutable_db_options_.stats_dump_period_sec;
-    if (stats_dump_period_sec > 0) {
-      if (!thread_dump_stats_) {
-        // In case of many `DB::Open()` in rapid succession we can have all
-        // threads dumping at once, which causes severe lock contention in
-        // jemalloc. Ensure successive `DB::Open()`s are staggered by at least
-        // one second in the common case.
-        static std::atomic<uint64_t> stats_dump_threads_started(0);
-        thread_dump_stats_.reset(new ROCKSDB_NAMESPACE::RepeatableThread(
-            [this]() { DBImpl::DumpStats(); }, "dump_st", env_,
-            static_cast<uint64_t>(stats_dump_period_sec) * kMicrosInSecond,
-            stats_dump_threads_started.fetch_add(1) %
-                static_cast<uint64_t>(stats_dump_period_sec) *
-                kMicrosInSecond));
-      }
-    }
-    stats_persist_period_sec = mutable_db_options_.stats_persist_period_sec;
-    if (stats_persist_period_sec > 0) {
-      if (!thread_persist_stats_) {
-        thread_persist_stats_.reset(new ROCKSDB_NAMESPACE::RepeatableThread(
-            [this]() { DBImpl::PersistStats(); }, "pst_st", env_,
-            static_cast<uint64_t>(stats_persist_period_sec) * kMicrosInSecond));
-      }
-    }
+    stats_dump_scheduler_ = StatsDumpScheduler::Default();
+#ifndef NDEBUG
+    stats_dump_scheduler_->TEST_UpdateEnv(env_);
```

Comment:
`TEST_Update()` is not thread-safe. Since the scheduler object can be shared by multiple `DBImpl`, using one DBImpl::mutex_ won't provide thread-safety.

Reply:
yes, you're right. I think we can add a mutex in `StatsDumpScheduler`, only for DEBUG.

Reply:
I wonder whether it's possible to call this function only for tests that use `MockEnv`.

Nit: maybe rename to `TEST_SetEnv`.

Reply:
Renamed to `TEST_SetEnv`.

It's tricky to only set `MockEnv` (or `SafeMockTimeEnv` here).
The test opens the DB with `SpecialEnv` first and then it reopens it with `SafeMockTimeEnv` (with setting options.env). As each test has it's own local `SafeMockTimeEnv`, vs. the static scheduler is shared between unittests. If we don't SetEnv for the first `SpicialEnv`, it will still use `SafeMockTimeEnv` from the last unittest, which is a local variable and destroyed.

If we could make env_ in timer non-const. We can add a test interface to update env in timer (from the unittest). Instead of right now, we have to provide the env_ during DB.open(). But I feel it would be better just have env_ in timer const (so it won't be changed while running).